opal/mca/threads/wait_sync.h

```diff
@@ -22,13 +30,109 @@
+#include "opal/runtime/opal_progress.h"
 #include "opal/sys/atomic.h"
 
 BEGIN_C_DECLS
 
 extern int opal_max_thread_in_progress;
 
-#include MCA_threads_wait_sync_base_include_HEADER
+typedef struct ompi_wait_sync_t {
+    opal_atomic_int32_t count;
+    int32_t status;
+    opal_thread_internal_cond_t condition;
+    opal_thread_internal_mutex_t lock;
+    struct ompi_wait_sync_t *next;
+    struct ompi_wait_sync_t *prev;
+    volatile bool signaling;
+} ompi_wait_sync_t;
+
+#define SYNC_WAIT(sync) (opal_using_threads() ? ompi_sync_wait_mt(sync) : sync_wait_st(sync))
+
+/* The loop in release handles a race condition between the signaling
+ * thread and the destruction of the condition variable. The signaling
+ * member will be set to false after the final signaling thread has
+ * finished operating on the sync object. This is done to avoid
+ * extra atomics in the signalling function and keep it as fast
+ * as possible. Note that the race window is small so spinning here
+ * is more optimal than sleeping since this macro is called in
+ * the critical path. */
+#define WAIT_SYNC_RELEASE(sync)                                \
+    if (opal_using_threads()) {                                \
+        while ((sync)->signaling) {                            \
+            if (opal_progress_yield_when_idle) {               \
+                opal_thread_yield();                           \
+            }                                                  \
+            continue;                                          \
+        }                                                      \
+        opal_thread_internal_cond_destroy(&(sync)->condition); \
+        opal_thread_internal_mutex_destroy(&(sync)->lock);     \
+    }
+
+#define WAIT_SYNC_RELEASE_NOWAIT(sync)                         \
+    if (opal_using_threads()) {                                \
+        opal_thread_internal_cond_destroy(&(sync)->condition); \
+        opal_thread_internal_mutex_destroy(&(sync)->lock);     \
+    }
+
+#define WAIT_SYNC_SIGNAL(sync)                                \
+    if (opal_using_threads()) {                               \
+        opal_thread_internal_mutex_lock(&(sync)->lock);       \
+        opal_thread_internal_cond_signal(&(sync)->condition); \
+        opal_thread_internal_mutex_unlock(&(sync)->lock);     \
+        (sync)->signaling = false;                            \
+    }
+
+#define WAIT_SYNC_SIGNALLED(sync)  \
+    {                              \
+        (sync)->signaling = false; \
+    }
+
+/* not static for inline "wait_sync_st" */
+OPAL_DECLSPEC extern ompi_wait_sync_t *wait_sync_list;
+
+OPAL_DECLSPEC int ompi_sync_wait_mt(ompi_wait_sync_t *sync);
+static inline int sync_wait_st(ompi_wait_sync_t *sync)
+{
+    assert(NULL == wait_sync_list);
+    assert(NULL == sync->next);
+    wait_sync_list = sync;
+
+    while (sync->count > 0) {
+        opal_progress();
+        if (opal_progress_yield_when_idle) {
```

Comment:
Thinking about this again: `opal_progress` will do the same check for `opal_progress_yield_when_idle` and yield if nothing has progressed. That will inevitably happen at some point if we're waiting for another ULT. I think we can get rid of this block altogether and rely on `opal_progress` to do yield if necessary.

Reply:
Yes you're right. It seems so: https://github.com/open-mpi/ompi/blob/master/opal/runtime/opal_progress.c#L216-L253

I will remove these branches in the next push.

Reply:
Thanks, I think this is good to go then. Can you please add a comment so that future us remembers why there is no yield?

Reply:
I removed `opal_thread_yield()` after `opal_progress()`. Please check the latest one.

Reply:
Much cleaner indeed. If you want a more precise yield (one that is not dependent on the network activity) you could register your own progress function, either high or low priority, and yield inside.

Reply:
@bosilca Thanks. I believe you are mentioning the following yield.
```c
#define WAIT_SYNC_RELEASE(sync)                                \
    if (opal_using_threads()) {                                \
        while ((sync)->signaling) {                            \
            if (opal_progress_yield_when_idle) {               \
                opal_thread_yield();                           \
            }                                                  \
            continue;                                          \
        }                                                      \
        opal_thread_internal_cond_destroy(&(sync)->condition); \
        opal_thread_internal_mutex_destroy(&(sync)->lock);     \
```
To be honest, I don't really know if this yield is really required for correctness when we use Qthreads or Argobots (it depends on the synchronization algorithm), but I'd like to keep it if I cannot confirm. If possible I'd be really happy if we could merge this PR once since this PR has already changed pretty many lines and logics (some are nontrivial).

Note that, even for Pthreads, it is not a bad idea to yield here since it should be better than busy wait (especially when the user sets `opal_progress_yield_when_idle=true` because of possible oversubscription or something).

Reply:
I was absolutely not mentioning this yield. This one is for releasing the memory pressure in high contention cases, and it should remain for performance reasons.

I was making a reference to our progress mechanism as provided by opal_progress_register and opal_progress_register_lp in opal/runtime/opal_progress.h.

Reply:
Oh I got it. Thanks for pointing it out! Yes, `opal_thread_yield()` in `opal_progress()` should have room for improvement. I believe that optimization is out of the scope of this PR, but I should be aware of this.